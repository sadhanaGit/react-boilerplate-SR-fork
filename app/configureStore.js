/**
 * Creates store
 * Creates reducer
 * Creates saga-middleware
 */

import { createStore, applyMiddleware, compose } from 'redux';
import { routerMiddleware } from 'connected-react-router';
import createSagaMiddleware from 'redux-saga';
import createReducer from './reducers';
import globalSaga from './containers/App/saga';

export default function configureStore(initialState = {}, history) {
  const composeEnhancers = compose;

  const sagaMiddleware = createSagaMiddleware();

  // Create the store with two middlewares
  // 1. sagaMiddleware: Makes redux-sagas work
  // 2. routerMiddleware: Syncs the location/URL path to the state
  const middlewares = [sagaMiddleware, routerMiddleware(history)];

  const enhancers = [applyMiddleware(...middlewares)];

  const store = createStore(
    createReducer(),
    initialState,
    composeEnhancers(...enhancers),
  );

  sagaMiddleware.run(globalSaga);

  return store;
}
