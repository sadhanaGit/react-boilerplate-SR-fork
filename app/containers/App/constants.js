/*
 * AppConstants
 * Each action has a corresponding type, which the reducer knows and picks up on.
 * To avoid weird typos between the reducer and the actions, we save them as
 * constants here. We prefix them with 'yourproject/YourComponent' so we avoid
 * reducers accidentally picking up actions they shouldn't.
 *
 * Follow this format:
 * export const YOUR_ACTION_CONSTANT = 'yourproject/YourContainer/YOUR_ACTION_CONSTANT';
 */

export const LOAD_STR = 'REACT-BOILERPLATE/App/LOAD_REPOS';
export const LOAD_STR_SUCCESS = 'REACT-BOILERPLATE/App/LOAD_STR_SUCCESS';
export const LOAD_STR_ERROR = 'REACT-BOILERPLATE/App/LOAD_STR_ERROR';
export const ADD_NEW_STR = 'REACT-BOILERPLATE/App/ADD_NEW_STR';
export const CHANGE_STRING = 'REACT-BOILERPLATE/App/CHANGE_STRING';
