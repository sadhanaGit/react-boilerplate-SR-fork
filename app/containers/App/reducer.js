/*
 * AppReducer
 *
 * The reducer takes care of our data. Using actions, we can
 * update our application state. To add a new action,
 * add it to the switch statement in the reducer function
 *
 */

// produce function of Immer library simplifies
// the process of writing immutable update logic, produce helps to avoid unknowingly change the state; takes two arguments first the current state and callback function,
import produce from 'immer';
import {
  LOAD_STR_SUCCESS,
  LOAD_STR,
  LOAD_STR_ERROR,
  ADD_NEW_STR,
  CHANGE_STRING,
} from './constants';

// The initial state of the App
export const initialState = {
  loading: false,
  error: false,
  newStr: '',
  strData: {
    repositories: false,
  },
};

/* eslint-disable default-case, no-param-reassign */
const appReducer = (state = initialState, action) =>
  produce(state, draft => {
    switch (action.type) {
      case LOAD_STR:
        draft.loading = true;
        draft.error = false;
        draft.strData.repositories = false;
        break;

      case LOAD_STR_SUCCESS:
        draft.strData.repositories = action.strlist;
        draft.loading = false;
        break;

      case LOAD_STR_ERROR:
        draft.error = action.error;
        draft.loading = false;
        break;

      case ADD_NEW_STR:
        draft.strData.repositories = [
          {
            id:
              Math.max(...state.strData.repositories.map(item => item.id)) + 1,
            text: state.newStr,
          },
          ...state.strData.repositories,
        ];
        draft.newStr = '';
        break;

      case CHANGE_STRING:
        draft.newStr = action.newStr;
        break;
    }
  });

export default appReducer;
