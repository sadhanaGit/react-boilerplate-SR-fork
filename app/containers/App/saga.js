/**
 * Gets the repositories of the user from Github
 */

import { call, put, takeLatest } from 'redux-saga/effects';
import { LOAD_STR } from 'containers/App/constants';
import { stringsLoaded, stringLoadingError } from 'containers/App/actions';

import request from 'utils/request';

/**
 * Github repos request/response handler
 */
export function* getRepos() {
  const requestURL = `http://localhost:3000/api/strlist`;

  try {
    // Call our request helper (see 'utils/request')
    const strlist = yield call(request, requestURL);
    yield put(stringsLoaded(strlist));
  } catch (err) {
    yield put(stringLoadingError(err));
  }
}

/**
 * Root saga manages watcher lifecycle
 */
export default function* stringListData() {
  // Watches for LOAD_STR actions and calls getRepos when one comes in.
  // By using `takeLatest` only the result of the latest API call is applied.
  // It returns task descriptor (just like fork) so we can continue execution
  // It will be cancelled automatically on component unmount
  yield takeLatest(LOAD_STR, getRepos);
}
