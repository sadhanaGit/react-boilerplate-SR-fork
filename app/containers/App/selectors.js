/**
 * The global state selectors
 */

import { createSelector } from 'reselect';
import { initialState } from './reducer';

const selectGlobal = state => state.global || initialState;

const selectRouter = state => state.router;

const makeSelectNewString = () =>
  createSelector(
    selectGlobal,
    globalState => globalState.newStr,
  );

const makeSelectLoading = () =>
  createSelector(
    selectGlobal,
    globalState => globalState.loading,
  );

const makeSelectError = () =>
  createSelector(
    selectGlobal,
    globalState => globalState.error,
  );

const makeSelectRepos = () =>
  createSelector(
    selectGlobal,
    globalState => globalState.strData.repositories,
  );

const makeSelectLocation = () =>
  createSelector(
    selectRouter,
    routerState => routerState.location,
  );

export {
  selectGlobal,
  makeSelectNewString,
  makeSelectLoading,
  makeSelectError,
  makeSelectRepos,
  makeSelectLocation,
};
