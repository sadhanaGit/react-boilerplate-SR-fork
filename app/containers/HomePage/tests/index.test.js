/**
 * Test the HomePage
 */

import React from 'react';
import { render } from 'react-testing-library';
import { Provider } from 'react-redux';
import { browserHistory } from 'react-router-dom';
import { HomePage, mapDispatchToProps } from '../index';
import { loadStrings } from '../../App/actions';
import configureStore from '../../../configureStore';

describe('<HomePage />', () => {
  let store;

  beforeAll(() => {
    store = configureStore({}, browserHistory);
  });

  it('should render and match the snapshot', () => {
    const {
      container: { firstChild },
    } = render(
      <Provider store={store}>
        <HomePage loading={false} error={false} repos={[]} />
      </Provider>,
    );
    expect(firstChild).toMatchSnapshot();
  });

  it('should fetch the repos on mount if a repos == false', () => {
    const submitSpy = jest.fn();
    render(
      <Provider store={store}>
        <HomePage
          repos={false}
          onLoadStringList={submitSpy}
        />
      </Provider>,
    );
    expect(submitSpy).toHaveBeenCalled();
  });

  it('should not call onLoadStringList if repos != false', () => {
    const submitSpy = jest.fn();
    render(
      <Provider store={store}>
        <HomePage repos="" onLoadStringList={submitSpy} />
      </Provider>,
    );
    expect(submitSpy).not.toHaveBeenCalled();
  });

  describe('onLoadStringList', () => {
    it('should be injected', () => {
      const dispatch = jest.fn();
      const result = mapDispatchToProps(dispatch);
      expect(result.onLoadStringList).toBeDefined();
    });

    it('should dispatch loadStrings when called', () => {
      const dispatch = jest.fn();
      const result = mapDispatchToProps(dispatch);
      result.onLoadStringList();
      expect(dispatch).toHaveBeenCalledWith(loadStrings());
    });

    it('should preventDefault if called with event', () => {
      const preventDefault = jest.fn();
      const result = mapDispatchToProps(() => { });
      const evt = { preventDefault };
      result.onLoadStringList(evt);
      expect(preventDefault).toHaveBeenCalledWith();
    });
  });
});
