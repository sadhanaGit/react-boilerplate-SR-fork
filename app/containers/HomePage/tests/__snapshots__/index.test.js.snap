// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HomePage /> should render and match the snapshot 1`] = `
<article>
  <div>
    <section
      class="Section-sc-1eulrsm-0 CenteredSection-sc-6zy0wx-0 imGCbL"
    >
      <h2
        class="H2-egphza-0 HsCXl"
      />
      <p />
    </section>
    <section
      class="Section-sc-1eulrsm-0 eTEsZl"
    >
      <b>
        String List
      </b>
      <p>
        List of strings from '/db/json.data' file on server via          API call
      </p>
      <div
        class="Wrapper-sc-1umgotm-0 laCwDT"
      >
        <ul
          class="Ul-mwnq6h-0 jcxXFy"
        />
      </div>
    </section>
  </div>
</article>
`;
